### ISSUE TITLE
[Common] Spec inconsistency with proto definition of Attributes

### ISSUE THREAD

**Issue by MSNev** [open-telemetry/opentelemetry-specification#2581]
The proto definition used for attributes uses "KeyValue" which can have a value of AnyValue, however, the spec definition of "Attribute" only allows for primitives or an Array of primitives, this is inconsistent.

What did you expect to see?
- They should be defined in the same way or have their own distinct definitions

**Additional context.**

The [Spec Attribute definition](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute)

Trace [proto attribute definition](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L225-L237)
```protobuf
  // attributes is a collection of key/value pairs. Note, global attributes
  // like server name can be set using the resource API. Examples of attributes:
  //
  //     "/http/user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
  //     "/http/server_latency": 300
  //     "abc.com/myattribute": true
  //     "abc.com/score": 10.239
  //
  // The OpenTelemetry API specification further restricts the allowed value types:
  // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes
  // Attribute keys MUST be unique (it is not allowed to have more than one
  // attribute with the same key).
  repeated opentelemetry.proto.common.v1.KeyValue attributes = 9;
```

And the [Logs spec also uses Attributes](https://github.com/open-telemetry/opentelemetry-specification/issues/2581)

Logs [proto attribute definition](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto#L198-L201)
```protobuf
  // Additional attributes that describe the specific event occurrence. [Optional].
  // Attribute keys MUST be unique (it is not allowed to have more than one
  // attribute with the same key).
  repeated opentelemetry.proto.common.v1.KeyValue attributes = 6;
```

With [Proto KeyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L61-L66) defined as taking AnyValue
```protobuf
// KeyValue is a key-value pair that is used to store Span attributes, Link
// attributes, etc.
message KeyValue {
  string key = 1;
  AnyValue value = 2;
}
```

And [Proto AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L24-L39) being 

```protobuf
// AnyValue is used to represent any type of attribute value. AnyValue may contain a
// primitive value such as a string or integer or it may contain an arbitrary nested
// object containing arrays, key-value lists and primitives.
message AnyValue {
  // The value is one of the listed fields. It is valid for all values to be unspecified
  // in which case this AnyValue is considered to be "empty".
  oneof value {
    string string_value = 1;
    bool bool_value = 2;
    int64 int_value = 3;
    double double_value = 4;
    ArrayValue array_value = 5;
    KeyValueList kvlist_value = 6;
    bytes bytes_value = 7;
  }
}
```

And [Proto KeyValueList](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L54-L60) being
```protobuf
// KeyValueList is a list of KeyValue messages. We need KeyValueList as a message
// since `oneof` in AnyValue does not allow repeated fields. Everywhere else where we need
// a list of KeyValue messages (e.g. in Span) we use `repeated KeyValue` directly to
// avoid unnecessary extra wrapping (which slows down the protocol). The 2 approaches
// are semantically equivalent.
message KeyValueList {
  // A collection of key/value pairs of key-value pairs. The list may be empty (may
  // contain 0 elements).
  // The keys MUST be unique (it is not allowed to have more than one
  // value with the same key).
  repeated KeyValue values = 1;
}
```

**Comment by Oberon00** [comment]
This is intentional. Maybe we should add a note in the spec about this?

**Comment by MSNev** [comment]
As the protocol supports nested attributes, what is wrong with expanding "all" Attributes definition to support nested attributes.
And for any backends that don't support nested attributes (thinking exporters) this would allow them to flattern out the attribute names.

This would also allow for unifying attributes usage across the board and usage, especially in relation to the forthcoming EventsApi,  which could allow for have a single API for both logs and trace transport mechanisms

**Comment by scheler** [comment]
@MSNev can you list some pros and cons of JSON.stringify'ing those nested objects as strings?

**Comment by MSNev** [comment]
JSON.stringify() is not the specific driver for this spec issue. The issue is that the definitions are not consistent which means an Attribute is not always the same Attribute.

We should be treating all definitions of Attribute (supporting nested Attributes) as the same. And as protobuf is defined as supporting nested Attributes then this is "just" as spec issue (on the assumption that SDK's don't have this coded this way, but this should not cause any breaking changes as it's expanding the options.

The only potential issues will be for exporters that don't support nesting and that can be resolved by the specific exporter to flatten the attribute keys.

JSON.stringify() Side Note (for completeness)
Pros:
for JSON exporter is that using the built-in native serialization rather than creating coded versions is more efficient for CPU and having nested Attributes is more efficient for memory both pre and post serialization.
Cons:
JSON can be verbose, but this depends on the object(s) being serialized

**Comment by tigrannajaryan** [comment]
> The proto definition used for attributes uses "KeyValue" which can have a value of AnyValue, however, the spec definition of "Attribute" only allows for primitives or an Array of primitives, this is inconsistent.

As @Oberon00 mentioned, this is intentional. OTLP is _at least_ as expressive as the Otel spec, but necessarily _exactly_ as expressive as the spec. We allow OTLP to be more expressive because Otel API/SDK is not the only source of data that we want to represent in OTLP. Other data sources may require richer capabilities that OTLP does offer, in particular the ability to represent arbitrarily nested objects.

>the spec definition of "Attribute" only allows for primitives or an Array of primitives, this is inconsistent.

This definition of "Attributes" currently applies to traces and metrics only, i.e. to the Stable parts of the spec. It does NOT apply to experimental logs, which has [its own definition](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#definitions-used-in-this-document) of what an attributes are and which allows nested objects.

>As the protocol supports nested attributes, what is wrong with expanding "all" Attributes definition to support nested attributes.

This can be discussed separately. In the past there were objections to allowing this for traces and metrics.


So, in short the existing differences between traces/metrics and logs and between the spec and OTLP are known and intentional.

**Comment by scheler** [comment]
if LogRecord allows objects/key-value pairs for attribute values, then that is sufficient to address the requirements of RUM. @MSNev?

**Comment by MSNev** [comment]
@tigrannajaryan That is where the inconsistencies are as the Spec also [states here](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute-collections) 

> "[Resources](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md), Metrics [data points](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#metric-points), [Spans](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-attributes), Span [Events](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events), Span [Links](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#specifying-links) and [Log Records](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md) may contain a collection of attributes.

@scheler While I found this as part of investigating the possibilities for the EventsAPI, this inconsistency is not specific to the RUM requirements. For Example the JS Sdk, supports nested objects as per the "Attribute Collections" definition.

**Comment by ocelotl** [comment]
I asked Claude to review this issue and Claude suggests it be closed with this explanation:

> The issue tracked that the spec's Attribute allowed only primitives/arrays while OTLP proto KeyValue allowed nested AnyValue. The current specification/common/README.md now defines an Attribute value as AnyValue (primitive, array of AnyValue, or map<string,AnyValue>), aligning spec with proto; related design issue #376 (nested/map attribute values) is now CLOSED. The direct PR #2620 was closed unmerged, but alignment was achieved by later AnyValue work, so the tracked inconsistency no longer exists.

To guard against a mistaken suggestion, Claude then independently re-verified this conclusion:

> Current common/README.md defines Attribute value as AnyValue (incl. map<string,AnyValue> and deep nesting), matching proto KeyValue/AnyValue; gating design issue #376 is CLOSED, so the tracked spec-vs-proto inconsistency is resolved.

Claude evaluated this issue with medium confidence that it can be closed.

Claude has classified this issue as resolved.

The intention of this comment is to make it easier for a human being to review this issue before making a decision.

**Comment by svrnm** [comment]
Closing this issue as it seems to be addresed, if this is incorrect please reopen or create a new one stating the details that are missing.
